sphinx/builders/linkcheck.py

```diff
@@ -437,9 +437,8 @@ def _check_uri(self, uri: str, hyperlink: Hyperlink) -> tuple[str, str, int]:
             except HTTPError as err:
                 error_message = str(err)
 
-                # Unauthorised: the reference probably exists
                 if status_code == 401:
-                    return 'working', 'unauthorized', 0
+                    return 'broken', 'unauthorized', 0
```

Comment:
> I think that the 401 still makes sense if people do not want to expose credentials in their configuration and only assume that something exists (and could return a 401).

Hmm.. point understood, although I think that even when an URL has no linkcheck credentials configured (the "no fake credential" case), an HTTP 401 response should be considered a broken link, because it means that the link is unchecked.  This is probably the core of the disruptive/controversial angle I have on this.

If for some reason a user is unable to provide credentials for some hyperlinks included in a documentation set, but wants a success report from linkchecking despite that, I'd argue they should use `linkcheck_anchors_ignore` to skip the relevant websites.

I'm still thinking about the first point -- using `None` as a special marker for intentionally-empty credentials.  It seems a bit fragile and too-clever; I like simple and hard-to-misunderstand things, because I misunderstand a lot.

> It's more of a follow-up, but what about using an enumeration (or at least mark the possible string statuses we returm using `Literal`) for the `status` instead of using `'working'`, `'broken'`, etc?

:+1: 

> In `process_result`, we use if-case but using `match` instead + enumerations may be more elegant and clearer (also we wouldn't have an arbitrary status code being emitted).

Mostly agree, although we still support Py3.9 at the moment - so this could be a future enhancement?

Reply:
Fine by me using `linkcheck_anchors_ignore` although it's a bit counterintuitive in this case (like I don't really want to ignore).

> using None as a special marker for intentionally-empty credentials

It was an example. We could have another configuration value like `linkcheck_auth_bypass` and put what links we expect to have 401.

Concerning the match, I forgot it was introduced in 3.10.

---

Actually, when writing my comment, I thought about a much more elegant solution, namely, you could specify the HTTP response code to expect for specific links and treat them as "working".

Reply:
> Fine by me using `linkcheck_anchors_ignore` although it's a bit counterintuitive in this case (like I don't really want to ignore).

Let me try to make a convicing argument that it makes sense :)  (I could be wrong!  in which case that'll help too)

The HTTP 401 response is zero-information in terms of hyperlink validity.  It does confirm that the client should use auth if it wants to gain more-than-zero information, but that's the only feedback it provides -- and we hide that from users at the moment.

So we're making network requests that fail, and we're not informing the user about it.  I think we should start by informing the user -- and then they either choose to find auth to gain greater-than-zero info, or they ignore the links and reduce the outbound traffic (and linkchecking time).

Reply:
>  I think we should start by informing the user 

Agreed. Actually, by counterintuitive, I meant that using the `linkcheck_anchors_ignore` was weird to me since I would have used it to suppress errors related to anchors rather than to HTTP response codes.

So, we could keep what you suggested.

Reply:
Agh, sorry.  I should have written `linkcheck_ignore` instead.  I completely failed to notice/parse the word `anchor` in there.

Reply:
Now it's ok for me !

Reply:
:)  thank you!

Reply:
>     * It's more of a follow-up, but what about using an enumeration (or at least mark the possible string statuses we returm using `Literal`) for the `status` instead of using `'working'`, `'broken'`, etc?
>       In `process_result`, we use if-case but   using `match` instead + enumerations may be more elegant and clearer (also we wouldn't have an arbitrary status code being emitted).

I'm working on some refactoring for this at the moment; the `CheckResult` class relies upon being JSON-serializable at the moment, and I don't feel like introducing custom serialization code, so I've opted to use the `StrEnum` type -- `Enum` with string values isn't serializable by default.

Taking that approach adds a dependency on Py3.11, so I'm experimenting with adding `match` statements at the same time.

I did read up a little bit about using `Literal` - it looks like it's mostly a type-checking aid?  I _think_ I'd prefer the enum route, even if it may take a while before it could land in the codebase.  Maybe I'll change my mind as I learn more, though..